# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [1]:
# Your code goes here
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

import cleanfunctions

In [2]:
# load and check file1.csv
df1 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')
display(df1)

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [3]:
# load and check file2.csv
df2 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv')
display(df2)
df2.isnull().sum()

Customer          ST GENDER Education Customer Lifetime Value  Income  \
0    GS98873     Arizona      F  Bachelor              323912.47%   16061   
1    CW49887  California      F    Master              462680.11%   79487   
2    MY31220  California      F   College              899704.02%   54230   
3    UH35128      Oregon      F   College             2580706.30%   71210   
4    WH52799     Arizona      F   College              380812.21%   94903   
..       ...         ...    ...       ...                     ...     ...   
991  HV85198     Arizona      M    Master              847141.75%   63513   
992  BS91566     Arizona      F   College              543121.91%   58161   
993  IL40123      Nevada      F   College              568964.41%   83640   
994  MY32149  California      F    Master              368672.38%       0   
995  SA91515  California      M  Bachelor              399258.39%       0   

     Monthly Premium Auto Number of Open Complaints  Total Claim Amount  \
0                      88                    1/0/00          633.600000   
1                     114                    1/0/00          547.200000   
2                     112                    1/0/00          537.600000   
3                     214                    1/1/00         1027.200000   
4                      94                    1/0/00          451.200000   
..                    ...                       ...                 ...   
991                    70                    1/0/00          185.667213   
992                    68                    1/0/00          140.747286   
993                    70                    1/0/00          471.050488   
994                    96                    1/0/00           28.460568   
995                   111                    1/0/00          700.349052   

        Policy Type  Vehicle Class  
0     Personal Auto  Four-Door Car  
1      Special Auto            SUV  
2     Personal Auto   Two-Door Car  
3     Personal Auto     Luxury Car  
4    Corporate Auto   Two-Door Car  
..              ...            ...  
991   Personal Auto  Four-Door Car  
992  Corporate Auto  Four-Door Car  
993  Corporate Auto   Two-Door Car  
994   Personal Auto   Two-Door Car  
995   Personal Auto            SUV  

[996 rows x 11 columns]

Customer                     0
ST                           0
GENDER                       5
Education                    0
Customer Lifetime Value      4
Income                       0
Monthly Premium Auto         0
Number of Open Complaints    0
Total Claim Amount           0
Policy Type                  0
Vehicle Class                0
dtype: int64

In [4]:
# load and check file3.csv
df3 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv')
display(df3)

Customer       State  Customer Lifetime Value             Education  \
0     SA25987  Washington              3479.137523  High School or Below   
1     TB86706     Arizona              2502.637401                Master   
2     ZL73902      Nevada              3265.156348              Bachelor   
3     KX23516  California              4455.843406  High School or Below   
4     FN77294  California              7704.958480  High School or Below   
...       ...         ...                      ...                   ...   
7065  LA72316  California             23405.987980              Bachelor   
7066  PK87824  California              3096.511217               College   
7067  TD14365  California              8163.890428              Bachelor   
7068  UP19263  California              7524.442436               College   
7069  Y167826  California              2611.836866               College   

     Gender  Income  Monthly Premium Auto  Number of Open Complaints  \
0         M       0                   104                          0   
1         M       0                    66                          0   
2         F   25820                    82                          0   
3         F       0                   121                          0   
4         M   30366                   101                          2   
...     ...     ...                   ...                        ...   
7065      M   71941                    73                          0   
7066      F   21604                    79                          0   
7067      M       0                    85                          3   
7068      M   21941                    96                          0   
7069      M       0                    77                          0   

         Policy Type  Total Claim Amount  Vehicle Class  
0      Personal Auto          499.200000   Two-Door Car  
1      Personal Auto            3.468912   Two-Door Car  
2      Personal Auto          393.600000  Four-Door Car  
3      Personal Auto          699.615192            SUV  
4      Personal Auto          484.800000            SUV  
...              ...                 ...            ...  
7065   Personal Auto          198.234764  Four-Door Car  
7066  Corporate Auto          379.200000  Four-Door Car  
7067  Corporate Auto          790.784983  Four-Door Car  
7068   Personal Auto          691.200000  Four-Door Car  
7069  Corporate Auto          369.600000   Two-Door Car  

[7070 rows x 11 columns]

In [5]:
# checking dataframes's info
display(df1.info())
display(df2.info())
display(df3.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 996 entries, 0 to 995
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   996 non-null    object 
 1   ST                         996 non-null    object 
 2   GENDER                     991 non-null    object 
 3   Education                  996 non-null    object 
 4   Customer Lifetime Value    992 non-null    object 
 5   Income                     996 non-null    int64  
 6   Monthly Premium Auto       996 non-null    int64  
 7   Number of Open Complaints  996 non-null    object 
 8   Total Claim Amount         996 non-null    float64
 9   Policy Type                996 non-null    object 
 10  Vehicle Class              996 non-null    object 
dtypes: float64(1), int64(2), object(8)
memory usage: 85.7+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7070 entries, 0 to 7069
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   7070 non-null   object 
 1   State                      7070 non-null   object 
 2   Customer Lifetime Value    7070 non-null   float64
 3   Education                  7070 non-null   object 
 4   Gender                     7070 non-null   object 
 5   Income                     7070 non-null   int64  
 6   Monthly Premium Auto       7070 non-null   int64  
 7   Number of Open Complaints  7070 non-null   int64  
 8   Policy Type                7070 non-null   object 
 9   Total Claim Amount         7070 non-null   float64
 10  Vehicle Class              7070 non-null   object 
dtypes: float64(2), int64(3), object(6)
memory usage: 607.7+ KB


None

# load cleaned file1 from previous lab

In [6]:
# read the cleaned file1 from previous lab
# path = "/Users/minhnguyen/IronHack2023-2024/Bootcamp/Week_1/Day_3/Afternoon/lab-dw-data-cleaning-and-formatting"

df1_clean = pd.read_csv("/Users/minhnguyen/IronHack2023-2024/Bootcamp/Week_1/Day_3/Afternoon/lab-dw-data-cleaning-and-formatting/customer_data_insurance_comp.csv")
df1_clean.head()
df1_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 952 entries, 0 to 951
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   customer                   952 non-null    object
 1   state                      952 non-null    object
 2   gender                     952 non-null    object
 3   education                  952 non-null    object
 4   customer_lifetime_value    952 non-null    int64 
 5   income                     952 non-null    int64 
 6   monthly_premium_auto       952 non-null    int64 
 7   number_of_open_complaints  952 non-null    int64 
 8   policy_type                952 non-null    object
 9   vehicle_class              952 non-null    object
 10  total_claim_amount         952 non-null    int64 
dtypes: int64(5), object(6)
memory usage: 81.9+ KB


In [7]:
# checking columns names of df1, df2, df3
display(df1_clean.columns)
display(df2.columns)
display(df3.columns)

# Insights:
# 1. first need to clean column name
# 2. re-positioning columns to match df1, df2, df3

Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')

Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Total Claim Amount', 'Policy Type', 'Vehicle Class'],
      dtype='object')

Index(['Customer', 'State', 'Customer Lifetime Value', 'Education', 'Gender',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Total Claim Amount', 'Vehicle Class'],
      dtype='object')

In [8]:
# convert column names to lower case
# apply cleanfunctions.lower_case_no_space_colname to df2, df3

df2_clean = cleanfunctions.lower_case_no_space_colname(df2)
df3_clean = cleanfunctions.lower_case_no_space_colname(df3)

In [9]:
# check df1, df2, df3 after clean_colname
display(df1_clean.columns)
display(df2_clean.columns)
display(df3_clean.columns)

# checking whether column names match
display(df1_clean.columns == df2_clean.columns)
display(df1_clean.columns == df3_clean.columns)


# manually correct column name 'st' to 'state' for df2

df2_clean = df2_clean.rename(columns = ({'st':'state'}))
display(df1_clean.columns)
display(df2_clean.columns)

Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')

Index(['customer', 'st', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'total_claim_amount', 'policy_type', 'vehicle_class'],
      dtype='object')

Index(['customer', 'state', 'customer_lifetime_value', 'education', 'gender',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'total_claim_amount', 'vehicle_class'],
      dtype='object')

array([ True, False,  True,  True,  True,  True,  True,  True, False,
       False, False])

array([ True,  True, False,  True, False,  True,  True,  True,  True,
       False, False])

Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')

Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'total_claim_amount', 'policy_type', 'vehicle_class'],
      dtype='object')

In [10]:
# re-positioning columns in df2, df3 to match df1

df2_clean = df2_clean[list(df1_clean.columns)]
display(df2_clean)

df3_clean = df3_clean[list(df1_clean.columns)]
display(df3_clean)


# final check for column name
display(df1_clean.columns == df2_clean.columns)
display(df1_clean.columns == df3_clean.columns)

customer       state gender education customer_lifetime_value  income  \
0    GS98873     Arizona      F  Bachelor              323912.47%   16061   
1    CW49887  California      F    Master              462680.11%   79487   
2    MY31220  California      F   College              899704.02%   54230   
3    UH35128      Oregon      F   College             2580706.30%   71210   
4    WH52799     Arizona      F   College              380812.21%   94903   
..       ...         ...    ...       ...                     ...     ...   
991  HV85198     Arizona      M    Master              847141.75%   63513   
992  BS91566     Arizona      F   College              543121.91%   58161   
993  IL40123      Nevada      F   College              568964.41%   83640   
994  MY32149  California      F    Master              368672.38%       0   
995  SA91515  California      M  Bachelor              399258.39%       0   

     monthly_premium_auto number_of_open_complaints     policy_type  \
0                      88                    1/0/00   Personal Auto   
1                     114                    1/0/00    Special Auto   
2                     112                    1/0/00   Personal Auto   
3                     214                    1/1/00   Personal Auto   
4                      94                    1/0/00  Corporate Auto   
..                    ...                       ...             ...   
991                    70                    1/0/00   Personal Auto   
992                    68                    1/0/00  Corporate Auto   
993                    70                    1/0/00  Corporate Auto   
994                    96                    1/0/00   Personal Auto   
995                   111                    1/0/00   Personal Auto   

     vehicle_class  total_claim_amount  
0    Four-Door Car          633.600000  
1              SUV          547.200000  
2     Two-Door Car          537.600000  
3       Luxury Car         1027.200000  
4     Two-Door Car          451.200000  
..             ...                 ...  
991  Four-Door Car          185.667213  
992  Four-Door Car          140.747286  
993   Two-Door Car          471.050488  
994   Two-Door Car           28.460568  
995            SUV          700.349052  

[996 rows x 11 columns]

customer       state gender             education  \
0     SA25987  Washington      M  High School or Below   
1     TB86706     Arizona      M                Master   
2     ZL73902      Nevada      F              Bachelor   
3     KX23516  California      F  High School or Below   
4     FN77294  California      M  High School or Below   
...       ...         ...    ...                   ...   
7065  LA72316  California      M              Bachelor   
7066  PK87824  California      F               College   
7067  TD14365  California      M              Bachelor   
7068  UP19263  California      M               College   
7069  Y167826  California      M               College   

      customer_lifetime_value  income  monthly_premium_auto  \
0                 3479.137523       0                   104   
1                 2502.637401       0                    66   
2                 3265.156348   25820                    82   
3                 4455.843406       0                   121   
4                 7704.958480   30366                   101   
...                       ...     ...                   ...   
7065             23405.987980   71941                    73   
7066              3096.511217   21604                    79   
7067              8163.890428       0                    85   
7068              7524.442436   21941                    96   
7069              2611.836866       0                    77   

      number_of_open_complaints     policy_type  vehicle_class  \
0                             0   Personal Auto   Two-Door Car   
1                             0   Personal Auto   Two-Door Car   
2                             0   Personal Auto  Four-Door Car   
3                             0   Personal Auto            SUV   
4                             2   Personal Auto            SUV   
...                         ...             ...            ...   
7065                          0   Personal Auto  Four-Door Car   
7066                          0  Corporate Auto  Four-Door Car   
7067                          3  Corporate Auto  Four-Door Car   
7068                          0   Personal Auto  Four-Door Car   
7069                          0  Corporate Auto   Two-Door Car   

      total_claim_amount  
0             499.200000  
1               3.468912  
2             393.600000  
3             699.615192  
4             484.800000  
...                  ...  
7065          198.234764  
7066          379.200000  
7067          790.784983  
7068          691.200000  
7069          369.600000  

[7070 rows x 11 columns]

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True])

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True])

# clean df2

In [11]:
# check df2_clean
display(df2_clean.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 996 entries, 0 to 995
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   996 non-null    object 
 1   state                      996 non-null    object 
 2   gender                     991 non-null    object 
 3   education                  996 non-null    object 
 4   customer_lifetime_value    992 non-null    object 
 5   income                     996 non-null    int64  
 6   monthly_premium_auto       996 non-null    int64  
 7   number_of_open_complaints  996 non-null    object 
 8   policy_type                996 non-null    object 
 9   vehicle_class              996 non-null    object 
 10  total_claim_amount         996 non-null    float64
dtypes: float64(1), int64(2), object(8)
memory usage: 85.7+ KB


None

In [12]:
# check value count for df2_clean
cleanfunctions.check_value_count_each_column(df2_clean)


Value counts of customer 
 GS98873    1
MI62456    1
SB50826    1
XX29952    1
AT93647    1
          ..
ZI47663    1
GO14008    1
BC43958    1
AI61368    1
SA91515    1
Name: customer, Length: 996, dtype: int64

Value counts of state 
 Oregon        303
California    277
Arizona       142
Nevada        125
Washington    100
AZ             49
Name: state, dtype: int64

Value counts of gender 
 F         527
M         461
female      2
Male        1
Name: gender, dtype: int64

Value counts of education 
 College                 311
High School or Below    290
Bachelor                243
Master                   86
Doctor                   49
Bachelors                17
Name: education, dtype: int64

Value counts of customer_lifetime_value 
 552821.28%     4
600521.48%     3
538949.95%     3
1203662.05%    3
871704.98%     3
              ..
728503.10%     1
1765196.34%    1
1702989.34%    1
403931.50%     1
399258.39%     1
Name: customer_lifetime_value, Length: 950, dtype: int64

Valu

df2_clean:
1. 'state': AZ to Arizona
2. 'gender': inconsitency annotation to F, M, U
3. 'education': Bachelors --> Bachelor
4. 'customer_lifetime_value': omit %, to float64
5. 'number_of_open_complaints': correct value, to int64
6. 'vehicle_class': to luxury
7. 'total_claim_amount': to int64

In [13]:
import cleanfunctions

# correct 'state' AZ to Arizona
df2_clean['state'] = df2_clean['state'].apply(lambda x: 'Arizona' if str(x) == 'AZ' else str(x))
df2_clean['state'].value_counts()

# clean gender column
df2_clean = cleanfunctions.clean_gender_column(df2_clean)
df2_clean['gender'].value_counts()


# correct 'education' Bachelors --> Bachelor
df2_clean['education'] = df2_clean['education'].apply(lambda x: 'Bachelor' if str(x) == 'Bachelors' else str(x))
df2_clean['education'].value_counts()

# clean 'customer_lifetime_value'
df2_clean['customer_lifetime_value'] = df2_clean['customer_lifetime_value'].apply(lambda x: str(x).replace('%','') if ('%' in str(x)) else str(x))
df2_clean['customer_lifetime_value'] = df2_clean['customer_lifetime_value'].astype('float64')

# clean 'number_of_open_complaints'
df2_clean['number_of_open_complaints'] = df2_clean['number_of_open_complaints'].apply(lambda x: int(str(x).split('/')[1] if pd.notnull(x) else x))
display(df2_clean['number_of_open_complaints'].value_counts()) # check counts and unique values after correcting

# clean 'vehicle_class': to luxury
df2_clean['vehicle_class'] = df2_clean['vehicle_class'].apply(lambda x: 'Luxury' if x in ['Sports Car', 'Luxury SUV', 'Luxury Car'] else x)
df2_clean['vehicle_class'].value_counts()

# based on df2_clean.info, there are 5 records with null value for
# column 'gender' which was converted to value 'U' as unknown
# and another 4 records with null value in column 'customer_lifetime_value'
# which will be dropped (as it is 4 records out of total 996 records)

# drop na
df2_clean = df2_clean.dropna()
df2_clean = df2_clean.reset_index()

# dop column 'index'
df2_clean = df2_clean.drop(columns='index')
df2_clean

# convert all numeric columns to integer
df2_clean = df2_clean.applymap(lambda x: int(x) if isinstance(x, (float,int)) else x)
display(df2_clean.dtypes)

                                    


0    796
1    109
2     43
3     26
4     16
5      6
Name: number_of_open_complaints, dtype: int64

customer                     object
state                        object
gender                       object
education                    object
customer_lifetime_value       int64
income                        int64
monthly_premium_auto          int64
number_of_open_complaints     int64
policy_type                  object
vehicle_class                object
total_claim_amount            int64
dtype: object

# clean df3

In [14]:
# check value count for df3_clean
cleanfunctions.check_value_count_each_column(df3_clean)


Value counts of customer 
 SA25987    1
EO99750    1
JM98365    1
QI11566    1
VG43444    1
          ..
TR96772    1
JU64043    1
ZE59592    1
FT74388    1
Y167826    1
Name: customer, Length: 7070, dtype: int64

Value counts of state 
 California    2544
Oregon        1978
Arizona       1302
Nevada         659
Washington     587
Name: state, dtype: int64

Value counts of gender 
 F    3576
M    3494
Name: gender, dtype: int64

Value counts of education 
 Bachelor                2152
College                 2058
High School or Below    2030
Master                   572
Doctor                   258
Name: education, dtype: int64

Value counts of customer_lifetime_value 
 2647.519741     6
2659.705650     6
4686.469674     6
4270.034394     6
13727.799720    6
               ..
11281.524510    1
8759.592518     1
4928.001408     1
4896.261605     1
2611.836866     1
Name: customer_lifetime_value, Length: 6288, dtype: int64

Value counts of income 
 0        1821
95697       8
74015     

In [15]:
df3_clean.info()
# df3 has no null value

# to clean for df3
# 1. column 'vehicle_class'
df3_clean['vehicle_class'] = df3_clean['vehicle_class'].apply(lambda x: 'Luxury' if x in ['Sports Car', 'Luxury SUV', 'Luxury Car'] else x)
df3_clean['vehicle_class'].value_counts()

# 2. cast all numeric columns to int64
df3_clean = df3_clean.applymap(lambda x: int(x) if isinstance(x, (float,int)) else x)
display(df3_clean.dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7070 entries, 0 to 7069
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   7070 non-null   object 
 1   state                      7070 non-null   object 
 2   gender                     7070 non-null   object 
 3   education                  7070 non-null   object 
 4   customer_lifetime_value    7070 non-null   float64
 5   income                     7070 non-null   int64  
 6   monthly_premium_auto       7070 non-null   int64  
 7   number_of_open_complaints  7070 non-null   int64  
 8   policy_type                7070 non-null   object 
 9   vehicle_class              7070 non-null   object 
 10  total_claim_amount         7070 non-null   float64
dtypes: float64(2), int64(3), object(6)
memory usage: 607.7+ KB


customer                     object
state                        object
gender                       object
education                    object
customer_lifetime_value       int64
income                        int64
monthly_premium_auto          int64
number_of_open_complaints     int64
policy_type                  object
vehicle_class                object
total_claim_amount            int64
dtype: object

In [16]:
# combining df1_clean, df2_clean, df3_clean into merged_df
merged_df = pd.concat([df1_clean, df2_clean, df3_clean], axis = 0)
merged_df

customer       state gender             education  \
0     QZ44356     Arizona      F              Bachelor   
1     AI49188      Nevada      F              Bachelor   
2     WW63253  California      M              Bachelor   
3     GA49547  Washington      M  High School or Below   
4     OC83172      Oregon      F              Bachelor   
...       ...         ...    ...                   ...   
7065  LA72316  California      M              Bachelor   
7066  PK87824  California      F               College   
7067  TD14365  California      M              Bachelor   
7068  UP19263  California      M               College   
7069  Y167826  California      M               College   

      customer_lifetime_value  income  monthly_premium_auto  \
0                      697953       0                    94   
1                     1288743   48767                   108   
2                      764586       0                   106   
3                      536307   36357                    68   
4                      825629   62902                    69   
...                       ...     ...                   ...   
7065                    23405   71941                    73   
7066                     3096   21604                    79   
7067                     8163       0                    85   
7068                     7524   21941                    96   
7069                     2611       0                    77   

      number_of_open_complaints     policy_type  vehicle_class  \
0                             0   Personal Auto  Four-Door Car   
1                             0   Personal Auto   Two-Door Car   
2                             0  Corporate Auto            SUV   
3                             0   Personal Auto  Four-Door Car   
4                             0   Personal Auto   Two-Door Car   
...                         ...             ...            ...   
7065                          0   Personal Auto  Four-Door Car   
7066                          0  Corporate Auto  Four-Door Car   
7067                          3  Corporate Auto  Four-Door Car   
7068                          0   Personal Auto  Four-Door Car   
7069                          0  Corporate Auto   Two-Door Car   

      total_claim_amount  
0                   1131  
1                    566  
2                    529  
3                     17  
4                    159  
...                  ...  
7065                 198  
7066                 379  
7067                 790  
7068                 691  
7069                 369  

[9014 rows x 11 columns]

In [17]:
# general check for na after concatenate dfs
merged_df.isnull().sum()

customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64

In [18]:
# check for duplicates in merged_df
merged_df.duplicated().sum()

# drop duplicates in merged_df
merged_df = merged_df.drop_duplicates()

# check after drop
merged_df.duplicated().sum()

0

In [19]:
# reset index of merged_df
merged_df = merged_df.reset_index()
#merged_df

#drop the index column
merged_df = merged_df.drop (columns = 'index')
merged_df

customer       state gender             education  \
0     QZ44356     Arizona      F              Bachelor   
1     AI49188      Nevada      F              Bachelor   
2     WW63253  California      M              Bachelor   
3     GA49547  Washington      M  High School or Below   
4     OC83172      Oregon      F              Bachelor   
...       ...         ...    ...                   ...   
9006  LA72316  California      M              Bachelor   
9007  PK87824  California      F               College   
9008  TD14365  California      M              Bachelor   
9009  UP19263  California      M               College   
9010  Y167826  California      M               College   

      customer_lifetime_value  income  monthly_premium_auto  \
0                      697953       0                    94   
1                     1288743   48767                   108   
2                      764586       0                   106   
3                      536307   36357                    68   
4                      825629   62902                    69   
...                       ...     ...                   ...   
9006                    23405   71941                    73   
9007                     3096   21604                    79   
9008                     8163       0                    85   
9009                     7524   21941                    96   
9010                     2611       0                    77   

      number_of_open_complaints     policy_type  vehicle_class  \
0                             0   Personal Auto  Four-Door Car   
1                             0   Personal Auto   Two-Door Car   
2                             0  Corporate Auto            SUV   
3                             0   Personal Auto  Four-Door Car   
4                             0   Personal Auto   Two-Door Car   
...                         ...             ...            ...   
9006                          0   Personal Auto  Four-Door Car   
9007                          0  Corporate Auto  Four-Door Car   
9008                          3  Corporate Auto  Four-Door Car   
9009                          0   Personal Auto  Four-Door Car   
9010                          0  Corporate Auto   Two-Door Car   

      total_claim_amount  
0                   1131  
1                    566  
2                    529  
3                     17  
4                    159  
...                  ...  
9006                 198  
9007                 379  
9008                 790  
9009                 691  
9010                 369  

[9011 rows x 11 columns]

In [20]:
# check value count for final merged_df
cleanfunctions.check_value_count_each_column(merged_df)


Value counts of customer 
 OB98149    2
PY42157    2
WQ89892    2
EE99484    2
KN13494    2
          ..
NI49607    1
MY11682    1
WW85018    1
TG75223    1
Y167826    1
Name: customer, Length: 8937, dtype: int64

Value counts of state 
 California    3112
Oregon        2557
Arizona       1680
Nevada         873
Washington     789
Name: state, dtype: int64

Value counts of gender 
 F    4602
M    4405
U       4
Name: gender, dtype: int64

Value counts of education 
 Bachelor                2701
College                 2647
High School or Below    2582
Master                   739
Doctor                   342
Name: education, dtype: int64

Value counts of customer_lifetime_value 
 6005     11
2619      9
2719      9
8382      9
3237      9
         ..
7713      1
21914     1
8071      1
8300      1
8163      1
Name: customer_lifetime_value, Length: 6552, dtype: int64

Value counts of income 
 0        2269
95697      13
10621       8
26876       8
61108       8
         ... 
62076     

In [21]:
# check info of merged_df
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9011 entries, 0 to 9010
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   customer                   9011 non-null   object
 1   state                      9011 non-null   object
 2   gender                     9011 non-null   object
 3   education                  9011 non-null   object
 4   customer_lifetime_value    9011 non-null   int64 
 5   income                     9011 non-null   int64 
 6   monthly_premium_auto       9011 non-null   int64 
 7   number_of_open_complaints  9011 non-null   int64 
 8   policy_type                9011 non-null   object
 9   vehicle_class              9011 non-null   object
 10  total_claim_amount         9011 non-null   int64 
dtypes: int64(5), object(6)
memory usage: 774.5+ KB


# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [22]:
# Your code goes here

#load and check marketing_customer_analysis.csv

marketing_customer_df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv')
marketing_customer_df


unnamed:_0 customer       state  customer_lifetime_value response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877  California              6857.519928       No   

       coverage education effective_to_date employmentstatus gender  ...  \
0         Basic   College        2011-02-18         Employed      M  ...   
1         Basic   College        2011-01-18       Unemployed      F  ...   
2         Basic  Bachelor        2011-02-10         Employed      M  ...   
3      Extended   College        2011-01-11         Employed      M  ...   
4       Premium  Bachelor        2011-01-17    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor        2011-01-19       Unemployed      F  ...   
10906     Basic   College        2011-01-06         Employed      F  ...   
10907  Extended  Bachelor        2011-02-06         Employed      F  ...   
10908   Premium   College        2011-02-13         Employed      F  ...   
10909     Basic  Bachelor        2011-01-08       Unemployed      M  ...   

       number_of_policies     policy_type        policy  renew_offer_type  \
0                       9  Corporate Auto  Corporate L3            Offer3   
1                       1   Personal Auto   Personal L3            Offer4   
2                       2   Personal Auto   Personal L3            Offer3   
3                       2  Corporate Auto  Corporate L3            Offer2   
4                       7   Personal Auto   Personal L2            Offer1   
...                   ...             ...           ...               ...   
10905                   7   Personal Auto   Personal L1            Offer3   
10906                   6   Personal Auto   Personal L3            Offer2   
10907                   2  Corporate Auto  Corporate L3            Offer1   
10908                   6   Personal Auto   Personal L1            Offer1   
10909                   3   Personal Auto   Personal L1            Offer4   

       sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
0              Agent          292.800000  Four-Door Car       Medsize   
1        Call Center          744.924331  Four-Door Car       Medsize   
2        Call Center          480.000000            SUV       Medsize   
3             Branch          484.013411  Four-Door Car       Medsize   
4             Branch          707.925645  Four-Door Car       Medsize   
...              ...                 ...            ...           ...   
10905            Web         1214.400000     Luxury Car       Medsize   
10906         Branch          273.018929  Four-Door Car       Medsize   
10907            Web          381.306996     Luxury SUV       Medsize   
10908         Branch          618.288849            SUV       Medsize   
10909            Web         1021.719397            SUV       Medsize   

      vehicle_type month  
0                A     2  
1                A     1  
2                A     2  
3                A     1  
4                A     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907            A     2  
10908            A     2  
10909            A     1  

[10910 rows x 27 columns]

In [23]:
# checking info about marketing_customer_df
marketing_customer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 27 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   unnamed:_0                     10910 non-null  int64  
 1   customer                       10910 non-null  object 
 2   state                          10910 non-null  object 
 3   customer_lifetime_value        10910 non-null  float64
 4   response                       10910 non-null  object 
 5   coverage                       10910 non-null  object 
 6   education                      10910 non-null  object 
 7   effective_to_date              10910 non-null  object 
 8   employmentstatus               10910 non-null  object 
 9   gender                         10910 non-null  object 
 10  income                         10910 non-null  int64  
 11  location_code                  10910 non-null  object 
 12  marital_status                 10910 non-null 

In [24]:
%%writefile -a '/Users/minhnguyen/IronHack2023-2024/Bootcamp/Week_1/Day_3/Afternoon/lab-dw-data-cleaning-and-formatting/cleanfunctions.py'

def check_value_count_each_column (df:pd.DataFrame):
    '''
    This function takes the input pandas dataframe, 
    then print value_counts for each column in the dataframe
    Input:
    df: Pandas dataframe
    Output:
    none
    '''
    for cols in list(df.columns):
        print('\nValue counts of',cols,'\n',df[cols].value_counts())



Appending to /Users/minhnguyen/IronHack2023-2024/Bootcamp/Week_1/Day_3/Afternoon/lab-dw-data-cleaning-and-formatting/cleanfunctions.py


In [25]:
cleanfunctions.check_value_count_each_column(marketing_customer_df)


Value counts of unnamed:_0 
 0        1
7276     1
7268     1
7269     1
7270     1
        ..
3638     1
3639     1
3640     1
3641     1
10909    1
Name: unnamed:_0, Length: 10910, dtype: int64

Value counts of customer 
 ID89933    7
DQ50258    5
AM79057    5
FP86246    5
KE35335    4
          ..
VW27730    1
ZX64745    1
CK24264    1
IX18485    1
WA60547    1
Name: customer, Length: 9134, dtype: int64

Value counts of state 
 California    4183
Oregon        2909
Arizona       1937
Nevada         993
Washington     888
Name: state, dtype: int64

Value counts of customer_lifetime_value 
 7507.455372     11
5504.139033     11
5181.620895     10
13727.799720    10
5332.462694     10
                ..
3467.323976      1
14140.558850     1
20018.204590     1
7432.299921      1
11971.977650     1
Name: customer_lifetime_value, Length: 8041, dtype: int64

Value counts of response 
 No     9444
Yes    1466
Name: response, dtype: int64

Value counts of coverage 
 Basic       6660
Extende

# checking for data type, value consistency:

# based on the dataset info and value_counts of each column

# these are my observations:

1. the dataset dimension is 10910 rows x 27 columns.Based on the info output, all columns has 10910 non-null count, indicating that this dataset has no null values

2. need to delete unnamed:_0 column

3. column 'effective_to_date' should be in datetime format, yyyy-mm-dd

4. column 'month_since_last_claim' (col[14]) --> dtype change to int64

5. column 'number_of_open_complaints' (col[16]) 
    * the current dtype is float64. This should be int64, not a float64 number as you can not have a half of a complaint.
    * considering the nature of the information of this column, the amount (633 records, is about 5.8% of total records),the approach I will take is to convert this colnumn to int64, and round the float value to the nearest int value.
  
6. change column name 'employmentstatus' to 'employment_status' for the consistency of column name

7. values in 'vehicle_class': 'Sport cars', 'luxury SUV', and 'luxury' could be replaced by 'luxury'

8. 'vehicle_size': replace 'medsize' with 'Medium'

9. 'vehicle_type': there is only 1 value of 'A'. it is not informative for the analysis of this dataset. However, it might be of used when this dataset is combined with another dataset with other values for 'vehicle_type' column. As it will not influence the analysis of the current dataset and might be of use in future analysis, there is no need to drop this column.

 

In [26]:
# check 16.'number_of_open_complaints', dtype float64 --> this should be int64 too
display(marketing_customer_df['number_of_open_complaints'].value_counts())

df_complaints_check = marketing_customer_df.loc[(marketing_customer_df['number_of_open_complaints']<1) & (marketing_customer_df['number_of_open_complaints'] >0)]
df_complaints_check

# why there is a large number of value '0.384256'
# judging by the type of data, this should be int64, not a float64 number
# as you can not have a half of a complaint
# considering the nature of the information of this column, the amount (5.8% of total records)
# the approach I will take is convert this colnumn to int64, and round the float value to the nearest int value.

0.000000    8160
1.000000    1145
0.384256     633
2.000000     414
3.000000     324
4.000000     166
5.000000      68
Name: number_of_open_complaints, dtype: int64

unnamed:_0 customer       state  customer_lifetime_value response  \
4               4  QA50777      Oregon              9025.067525       No   
23             23  NQ71171  California              5107.071054       No   
51             51  FT56968     Arizona              2590.096027       No   
59             59  EP83939     Arizona              5575.751228       No   
67             67  KR35099  Washington              7507.455372      Yes   
...           ...      ...         ...                      ...      ...   
10801       10801  AP56340  Washington              5384.671725       No   
10816       10816  AW71157  California              5531.723723       No   
10833       10833  OQ99704  Washington              3981.775889       No   
10834       10834  FU99476  Washington              6770.306834       No   
10905       10905  FE99816      Nevada             15563.369440       No   

       coverage             education effective_to_date employmentstatus  \
4       Premium              Bachelor        2011-01-17    Medical Leave   
23        Basic              Bachelor        2011-02-02         Employed   
51        Basic  High School or Below        2011-01-03         Employed   
59        Basic  High School or Below        2011-01-26         Employed   
67        Basic               College        2011-02-06         Employed   
...         ...                   ...               ...              ...   
10801     Basic              Bachelor        2011-02-26         Disabled   
10816     Basic               College        2011-01-16         Employed   
10833     Basic              Bachelor        2011-02-11         Employed   
10834  Extended  High School or Below        2011-02-22         Employed   
10905   Premium              Bachelor        2011-01-19       Unemployed   

      gender  ...  number_of_policies    policy_type       policy  \
4          F  ...                   7  Personal Auto  Personal L2   
23         M  ...                   7  Personal Auto  Personal L2   
51         M  ...                   1  Personal Auto  Personal L1   
59         M  ...                   5  Personal Auto  Personal L3   
67         M  ...                   2  Personal Auto  Personal L3   
...      ...  ...                 ...            ...          ...   
10801      M  ...                   7  Personal Auto  Personal L2   
10816      M  ...                   5  Personal Auto  Personal L1   
10833      F  ...                   1  Personal Auto  Personal L3   
10834      M  ...                   3  Personal Auto  Personal L1   
10905      F  ...                   7  Personal Auto  Personal L1   

       renew_offer_type  sales_channel  total_claim_amount  vehicle_class  \
4                Offer1         Branch          707.925645  Four-Door Car   
23               Offer1          Agent          128.900320  Four-Door Car   
51               Offer1          Agent          321.600000  Four-Door Car   
59               Offer2    Call Center          109.904496  Four-Door Car   
67               Offer2          Agent          231.201886   Two-Door Car   
...                 ...            ...                 ...            ...   
10801            Offer1          Agent          458.003651  Four-Door Car   
10816            Offer3            Web           17.484679  Four-Door Car   
10833            Offer2         Branch          475.200880            SUV   
10834            Offer1    Call Center           95.338505  Four-Door Car   
10905            Offer3            Web         1214.400000     Luxury Car   

       vehicle_size vehicle_type month  
4           Medsize            A     1  
23          Medsize            A     2  
51            Large            A     1  
59          Medsize            A     1  
67          Medsize            A     2  
...             ...          ...   ...  
10801       Medsize            A     2  
10816       Medsize            A     1  
10833         Small            A     2  
10834       Medsi

In [27]:
# although I mentioned that this dataset has no null values above
# but for the purpose of practice, I will check for null values with the .isnull code still
marketing_customer_df.isnull().sum()

# based on the output, there is no null values

unnamed:_0                       0
customer                         0
state                            0
customer_lifetime_value          0
response                         0
coverage                         0
education                        0
effective_to_date                0
employmentstatus                 0
gender                           0
income                           0
location_code                    0
marital_status                   0
monthly_premium_auto             0
months_since_last_claim          0
months_since_policy_inception    0
number_of_open_complaints        0
number_of_policies               0
policy_type                      0
policy                           0
renew_offer_type                 0
sales_channel                    0
total_claim_amount               0
vehicle_class                    0
vehicle_size                     0
vehicle_type                     0
month                            0
dtype: int64

In [28]:
# checking for duplicated records
marketing_customer_df.duplicated(subset=['months_since_last_claim','number_of_open_complaints']).sum()


# based on the output, there is no duplicated records

10701

In [29]:
# 2. need to delete unnamed:_0 column

marketing_customer_df_clean = marketing_customer_df.drop(columns='unnamed:_0')
marketing_customer_df_clean


customer       state  customer_lifetime_value response  coverage  \
0      DK49336     Arizona              4809.216960       No     Basic   
1      KX64629  California              2228.525238       No     Basic   
2      LZ68649  Washington             14947.917300       No     Basic   
3      XL78013      Oregon             22332.439460      Yes  Extended   
4      QA50777      Oregon              9025.067525       No   Premium   
...        ...         ...                      ...      ...       ...   
10905  FE99816      Nevada             15563.369440       No   Premium   
10906  KX53892      Oregon              5259.444853       No     Basic   
10907  TL39050     Arizona             23893.304100       No  Extended   
10908  WA60547  California             11971.977650       No   Premium   
10909  IV32877  California              6857.519928       No     Basic   

      education effective_to_date employmentstatus gender  income  ...  \
0       College        2011-02-18         Employed      M   48029  ...   
1       College        2011-01-18       Unemployed      F       0  ...   
2      Bachelor        2011-02-10         Employed      M   22139  ...   
3       College        2011-01-11         Employed      M   49078  ...   
4      Bachelor        2011-01-17    Medical Leave      F   23675  ...   
...         ...               ...              ...    ...     ...  ...   
10905  Bachelor        2011-01-19       Unemployed      F       0  ...   
10906   College        2011-01-06         Employed      F   61146  ...   
10907  Bachelor        2011-02-06         Employed      F   39837  ...   
10908   College        2011-02-13         Employed      F   64195  ...   
10909  Bachelor        2011-01-08       Unemployed      M       0  ...   

      number_of_policies     policy_type        policy  renew_offer_type  \
0                      9  Corporate Auto  Corporate L3            Offer3   
1                      1   Personal Auto   Personal L3            Offer4   
2                      2   Personal Auto   Personal L3            Offer3   
3                      2  Corporate Auto  Corporate L3            Offer2   
4                      7   Personal Auto   Personal L2            Offer1   
...                  ...             ...           ...               ...   
10905                  7   Personal Auto   Personal L1            Offer3   
10906                  6   Personal Auto   Personal L3            Offer2   
10907                  2  Corporate Auto  Corporate L3            Offer1   
10908                  6   Personal Auto   Personal L1            Offer1   
10909                  3   Personal Auto   Personal L1            Offer4   

       sales_channel  total_claim_amount  vehicle_class vehicle_size  \
0              Agent          292.800000  Four-Door Car      Medsize   
1        Call Center          744.924331  Four-Door Car      Medsize   
2        Call Center          480.000000            SUV      Medsize   
3             Branch          484.013411  Four-Door Car      Medsize   
4             Branch          707.925645  Four-Door Car      Medsize   
...              ...                 ...            ...          ...   
10905            Web         1214.400000     Luxury Car      Medsize   
10906         Branch          273.018929  Four-Door Car      Medsize   
10907            Web          381.306996     Luxury SUV      Medsize   
10908         Branch          618.288849            SUV      Medsize   
10909            Web         1021.719397            SUV      Medsize   

      vehicle_type month  
0                A     2  
1                A     1  
2                A     2  
3                A     1  
4                A     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907            A     2  
10908            A     2  
10909            A     1  

[10910 rows x 26 columns]

In [30]:
# 3. column 'effective_to_date' should be in datetime format, yyyy-mm-dd

marketing_customer_df_clean['effective_to_date'] = pd.to_datetime(marketing_customer_df['effective_to_date'], errors='coerce')
display(marketing_customer_df_clean)       

#print(marketing_customer_df.dtypes)
print(marketing_customer_df_clean.dtypes)

customer       state  customer_lifetime_value response  coverage  \
0      DK49336     Arizona              4809.216960       No     Basic   
1      KX64629  California              2228.525238       No     Basic   
2      LZ68649  Washington             14947.917300       No     Basic   
3      XL78013      Oregon             22332.439460      Yes  Extended   
4      QA50777      Oregon              9025.067525       No   Premium   
...        ...         ...                      ...      ...       ...   
10905  FE99816      Nevada             15563.369440       No   Premium   
10906  KX53892      Oregon              5259.444853       No     Basic   
10907  TL39050     Arizona             23893.304100       No  Extended   
10908  WA60547  California             11971.977650       No   Premium   
10909  IV32877  California              6857.519928       No     Basic   

      education effective_to_date employmentstatus gender  income  ...  \
0       College        2011-02-18         Employed      M   48029  ...   
1       College        2011-01-18       Unemployed      F       0  ...   
2      Bachelor        2011-02-10         Employed      M   22139  ...   
3       College        2011-01-11         Employed      M   49078  ...   
4      Bachelor        2011-01-17    Medical Leave      F   23675  ...   
...         ...               ...              ...    ...     ...  ...   
10905  Bachelor        2011-01-19       Unemployed      F       0  ...   
10906   College        2011-01-06         Employed      F   61146  ...   
10907  Bachelor        2011-02-06         Employed      F   39837  ...   
10908   College        2011-02-13         Employed      F   64195  ...   
10909  Bachelor        2011-01-08       Unemployed      M       0  ...   

      number_of_policies     policy_type        policy  renew_offer_type  \
0                      9  Corporate Auto  Corporate L3            Offer3   
1                      1   Personal Auto   Personal L3            Offer4   
2                      2   Personal Auto   Personal L3            Offer3   
3                      2  Corporate Auto  Corporate L3            Offer2   
4                      7   Personal Auto   Personal L2            Offer1   
...                  ...             ...           ...               ...   
10905                  7   Personal Auto   Personal L1            Offer3   
10906                  6   Personal Auto   Personal L3            Offer2   
10907                  2  Corporate Auto  Corporate L3            Offer1   
10908                  6   Personal Auto   Personal L1            Offer1   
10909                  3   Personal Auto   Personal L1            Offer4   

       sales_channel  total_claim_amount  vehicle_class vehicle_size  \
0              Agent          292.800000  Four-Door Car      Medsize   
1        Call Center          744.924331  Four-Door Car      Medsize   
2        Call Center          480.000000            SUV      Medsize   
3             Branch          484.013411  Four-Door Car      Medsize   
4             Branch          707.925645  Four-Door Car      Medsize   
...              ...                 ...            ...          ...   
10905            Web         1214.400000     Luxury Car      Medsize   
10906         Branch          273.018929  Four-Door Car      Medsize   
10907            Web          381.306996     Luxury SUV      Medsize   
10908         Branch          618.288849            SUV      Medsize   
10909            Web         1021.719397            SUV      Medsize   

      vehicle_type month  
0                A     2  
1                A     1  
2                A     2  
3                A     1  
4                A     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907            A     2  
10908            A     2  
10909            A     1  

[10910 rows x 26 columns]

customer                                 object
state                                    object
customer_lifetime_value                 float64
response                                 object
coverage                                 object
education                                object
effective_to_date                datetime64[ns]
employmentstatus                         object
gender                                   object
income                                    int64
location_code                            object
marital_status                           object
monthly_premium_auto                      int64
months_since_last_claim                 float64
months_since_policy_inception             int64
number_of_open_complaints               float64
number_of_policies                        int64
policy_type                              object
policy                                   object
renew_offer_type                         object
sales_channel                           

In [31]:
# 4. column 'months_since_last_claim' (col[14]) --> dtype change to int64
#marketing_customer_df['months_since_last_claim']= marketing_customer_df['months_since_last_claim'].applymap(lambda x:int(x) if isinstance (x, (float,int)) else x) 
marketing_customer_df_clean['months_since_last_claim']= marketing_customer_df['months_since_last_claim'].round().astype('int64') 

display(marketing_customer_df_clean['months_since_last_claim'].value_counts())
#display(type(marketing_customer_df['months_since_last_claim']))

#print(marketing_customer_df.dtypes)
print(marketing_customer_df_clean.dtypes)

display(marketing_customer_df_clean.head())

15    956
3     426
6     394
1     386
7     378
4     373
2     368
5     355
0     354
10    346
16    333
11    331
13    320
8     315
12    307
17    299
14    299
9     277
23    276
19    276
25    269
21    267
18    257
20    254
22    246
24    237
29    234
28    227
31    221
26    211
27    208
30    201
34    199
33    198
35    157
32    155
Name: months_since_last_claim, dtype: int64

customer                                 object
state                                    object
customer_lifetime_value                 float64
response                                 object
coverage                                 object
education                                object
effective_to_date                datetime64[ns]
employmentstatus                         object
gender                                   object
income                                    int64
location_code                            object
marital_status                           object
monthly_premium_auto                      int64
months_since_last_claim                   int64
months_since_policy_inception             int64
number_of_open_complaints               float64
number_of_policies                        int64
policy_type                              object
policy                                   object
renew_offer_type                         object
sales_channel                           

customer       state  customer_lifetime_value response  coverage education  \
0  DK49336     Arizona              4809.216960       No     Basic   College   
1  KX64629  California              2228.525238       No     Basic   College   
2  LZ68649  Washington             14947.917300       No     Basic  Bachelor   
3  XL78013      Oregon             22332.439460      Yes  Extended   College   
4  QA50777      Oregon              9025.067525       No   Premium  Bachelor   

  effective_to_date employmentstatus gender  income  ... number_of_policies  \
0        2011-02-18         Employed      M   48029  ...                  9   
1        2011-01-18       Unemployed      F       0  ...                  1   
2        2011-02-10         Employed      M   22139  ...                  2   
3        2011-01-11         Employed      M   49078  ...                  2   
4        2011-01-17    Medical Leave      F   23675  ...                  7   

      policy_type        policy  renew_offer_type  sales_channel  \
0  Corporate Auto  Corporate L3            Offer3          Agent   
1   Personal Auto   Personal L3            Offer4    Call Center   
2   Personal Auto   Personal L3            Offer3    Call Center   
3  Corporate Auto  Corporate L3            Offer2         Branch   
4   Personal Auto   Personal L2            Offer1         Branch   

   total_claim_amount  vehicle_class vehicle_size vehicle_type month  
0          292.800000  Four-Door Car      Medsize            A     2  
1          744.924331  Four-Door Car      Medsize            A     1  
2          480.000000            SUV      Medsize            A     2  
3          484.013411  Four-Door Car      Medsize            A     1  
4          707.925645  Four-Door Car      Medsize            A     1  

[5 rows x 26 columns]

In [32]:
#5. column 'number_of_open_complaints' (col[16]) 
#    * the current dtype is float64. This should be int64, not a float64 number as you can not have a half of a complaint.
#    * considering the nature of the information of this column, the amount (633 records, is about 5.8% of total records),the approach I will take is to convert this colnumn to int64, and round the float value to the nearest int value.
marketing_customer_df_clean['number_of_open_complaints']= marketing_customer_df['number_of_open_complaints'].round().astype('int64') 

display(marketing_customer_df_clean['number_of_open_complaints'].value_counts())

#print(marketing_customer_df.dtypes)
print(marketing_customer_df_clean.dtypes)
# Caution: after this data type convertion, the number of O complaints
# goes from 8160 to 8793 records. This might influence the interpretation of data.

display(marketing_customer_df_clean.head())

0    8793
1    1145
2     414
3     324
4     166
5      68
Name: number_of_open_complaints, dtype: int64

customer                                 object
state                                    object
customer_lifetime_value                 float64
response                                 object
coverage                                 object
education                                object
effective_to_date                datetime64[ns]
employmentstatus                         object
gender                                   object
income                                    int64
location_code                            object
marital_status                           object
monthly_premium_auto                      int64
months_since_last_claim                   int64
months_since_policy_inception             int64
number_of_open_complaints                 int64
number_of_policies                        int64
policy_type                              object
policy                                   object
renew_offer_type                         object
sales_channel                           

customer       state  customer_lifetime_value response  coverage education  \
0  DK49336     Arizona              4809.216960       No     Basic   College   
1  KX64629  California              2228.525238       No     Basic   College   
2  LZ68649  Washington             14947.917300       No     Basic  Bachelor   
3  XL78013      Oregon             22332.439460      Yes  Extended   College   
4  QA50777      Oregon              9025.067525       No   Premium  Bachelor   

  effective_to_date employmentstatus gender  income  ... number_of_policies  \
0        2011-02-18         Employed      M   48029  ...                  9   
1        2011-01-18       Unemployed      F       0  ...                  1   
2        2011-02-10         Employed      M   22139  ...                  2   
3        2011-01-11         Employed      M   49078  ...                  2   
4        2011-01-17    Medical Leave      F   23675  ...                  7   

      policy_type        policy  renew_offer_type  sales_channel  \
0  Corporate Auto  Corporate L3            Offer3          Agent   
1   Personal Auto   Personal L3            Offer4    Call Center   
2   Personal Auto   Personal L3            Offer3    Call Center   
3  Corporate Auto  Corporate L3            Offer2         Branch   
4   Personal Auto   Personal L2            Offer1         Branch   

   total_claim_amount  vehicle_class vehicle_size vehicle_type month  
0          292.800000  Four-Door Car      Medsize            A     2  
1          744.924331  Four-Door Car      Medsize            A     1  
2          480.000000            SUV      Medsize            A     2  
3          484.013411  Four-Door Car      Medsize            A     1  
4          707.925645  Four-Door Car      Medsize            A     1  

[5 rows x 26 columns]

In [33]:
#6. change column name 'employmentstatus' to 'employment_status' for the consistency of column name
marketing_customer_df_clean=marketing_customer_df_clean.rename(columns = {'employmentstatus': 'employment_status'})
marketing_customer_df_clean

customer       state  customer_lifetime_value response  coverage  \
0      DK49336     Arizona              4809.216960       No     Basic   
1      KX64629  California              2228.525238       No     Basic   
2      LZ68649  Washington             14947.917300       No     Basic   
3      XL78013      Oregon             22332.439460      Yes  Extended   
4      QA50777      Oregon              9025.067525       No   Premium   
...        ...         ...                      ...      ...       ...   
10905  FE99816      Nevada             15563.369440       No   Premium   
10906  KX53892      Oregon              5259.444853       No     Basic   
10907  TL39050     Arizona             23893.304100       No  Extended   
10908  WA60547  California             11971.977650       No   Premium   
10909  IV32877  California              6857.519928       No     Basic   

      education effective_to_date employment_status gender  income  ...  \
0       College        2011-02-18          Employed      M   48029  ...   
1       College        2011-01-18        Unemployed      F       0  ...   
2      Bachelor        2011-02-10          Employed      M   22139  ...   
3       College        2011-01-11          Employed      M   49078  ...   
4      Bachelor        2011-01-17     Medical Leave      F   23675  ...   
...         ...               ...               ...    ...     ...  ...   
10905  Bachelor        2011-01-19        Unemployed      F       0  ...   
10906   College        2011-01-06          Employed      F   61146  ...   
10907  Bachelor        2011-02-06          Employed      F   39837  ...   
10908   College        2011-02-13          Employed      F   64195  ...   
10909  Bachelor        2011-01-08        Unemployed      M       0  ...   

      number_of_policies     policy_type        policy  renew_offer_type  \
0                      9  Corporate Auto  Corporate L3            Offer3   
1                      1   Personal Auto   Personal L3            Offer4   
2                      2   Personal Auto   Personal L3            Offer3   
3                      2  Corporate Auto  Corporate L3            Offer2   
4                      7   Personal Auto   Personal L2            Offer1   
...                  ...             ...           ...               ...   
10905                  7   Personal Auto   Personal L1            Offer3   
10906                  6   Personal Auto   Personal L3            Offer2   
10907                  2  Corporate Auto  Corporate L3            Offer1   
10908                  6   Personal Auto   Personal L1            Offer1   
10909                  3   Personal Auto   Personal L1            Offer4   

       sales_channel  total_claim_amount  vehicle_class vehicle_size  \
0              Agent          292.800000  Four-Door Car      Medsize   
1        Call Center          744.924331  Four-Door Car      Medsize   
2        Call Center          480.000000            SUV      Medsize   
3             Branch          484.013411  Four-Door Car      Medsize   
4             Branch          707.925645  Four-Door Car      Medsize   
...              ...                 ...            ...          ...   
10905            Web         1214.400000     Luxury Car      Medsize   
10906         Branch          273.018929  Four-Door Car      Medsize   
10907            Web          381.306996     Luxury SUV      Medsize   
10908         Branch          618.288849            SUV      Medsize   
10909            Web         1021.719397            SUV      Medsize   

      vehicle_type month  
0                A     2  
1                A     1  
2                A     2  
3                A     1  
4                A     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907            A     2  
10908            A     2  
10909            A     1  

[10910 rows x 26 columns]

In [34]:
#7. values in 'vehicle_class': 'Sport cars', 'luxury SUV', and 'luxury' could be replaced by 'luxury'
marketing_customer_df_clean['vehicle_class']=marketing_customer_df_clean['vehicle_class'].apply(lambda x: 'Luxury' if x in ['Sports Car', 'Luxury SUV', 'Luxury Car'] else x) 
display(marketing_customer_df_clean['vehicle_class'].value_counts())

Four-Door Car    5834
Two-Door Car     2118
SUV              2012
Luxury            946
Name: vehicle_class, dtype: int64

In [35]:
#8. 'vehicle_size': replace 'medsize' with 'Medium'
marketing_customer_df_clean['vehicle_size'] = marketing_customer_df_clean['vehicle_size'].apply(lambda x: 'Medium' if x == 'Medsize' else x )
display(marketing_customer_df_clean['vehicle_size'].value_counts())

Medium    7873
Small     1966
Large     1071
Name: vehicle_size, dtype: int64

In [36]:
display(marketing_customer_df_clean)
display(marketing_customer_df_clean.dtypes)
display(marketing_customer_df_clean.info())

customer       state  customer_lifetime_value response  coverage  \
0      DK49336     Arizona              4809.216960       No     Basic   
1      KX64629  California              2228.525238       No     Basic   
2      LZ68649  Washington             14947.917300       No     Basic   
3      XL78013      Oregon             22332.439460      Yes  Extended   
4      QA50777      Oregon              9025.067525       No   Premium   
...        ...         ...                      ...      ...       ...   
10905  FE99816      Nevada             15563.369440       No   Premium   
10906  KX53892      Oregon              5259.444853       No     Basic   
10907  TL39050     Arizona             23893.304100       No  Extended   
10908  WA60547  California             11971.977650       No   Premium   
10909  IV32877  California              6857.519928       No     Basic   

      education effective_to_date employment_status gender  income  ...  \
0       College        2011-02-18          Employed      M   48029  ...   
1       College        2011-01-18        Unemployed      F       0  ...   
2      Bachelor        2011-02-10          Employed      M   22139  ...   
3       College        2011-01-11          Employed      M   49078  ...   
4      Bachelor        2011-01-17     Medical Leave      F   23675  ...   
...         ...               ...               ...    ...     ...  ...   
10905  Bachelor        2011-01-19        Unemployed      F       0  ...   
10906   College        2011-01-06          Employed      F   61146  ...   
10907  Bachelor        2011-02-06          Employed      F   39837  ...   
10908   College        2011-02-13          Employed      F   64195  ...   
10909  Bachelor        2011-01-08        Unemployed      M       0  ...   

      number_of_policies     policy_type        policy  renew_offer_type  \
0                      9  Corporate Auto  Corporate L3            Offer3   
1                      1   Personal Auto   Personal L3            Offer4   
2                      2   Personal Auto   Personal L3            Offer3   
3                      2  Corporate Auto  Corporate L3            Offer2   
4                      7   Personal Auto   Personal L2            Offer1   
...                  ...             ...           ...               ...   
10905                  7   Personal Auto   Personal L1            Offer3   
10906                  6   Personal Auto   Personal L3            Offer2   
10907                  2  Corporate Auto  Corporate L3            Offer1   
10908                  6   Personal Auto   Personal L1            Offer1   
10909                  3   Personal Auto   Personal L1            Offer4   

       sales_channel  total_claim_amount  vehicle_class vehicle_size  \
0              Agent          292.800000  Four-Door Car       Medium   
1        Call Center          744.924331  Four-Door Car       Medium   
2        Call Center          480.000000            SUV       Medium   
3             Branch          484.013411  Four-Door Car       Medium   
4             Branch          707.925645  Four-Door Car       Medium   
...              ...                 ...            ...          ...   
10905            Web         1214.400000         Luxury       Medium   
10906         Branch          273.018929  Four-Door Car       Medium   
10907            Web          381.306996         Luxury       Medium   
10908         Branch          618.288849            SUV       Medium   
10909            Web         1021.719397            SUV       Medium   

      vehicle_type month  
0                A     2  
1                A     1  
2                A     2  
3                A     1  
4                A     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907            A     2  
10908            A     2  
10909            A     1  

[10910 rows x 26 columns]

customer                                 object
state                                    object
customer_lifetime_value                 float64
response                                 object
coverage                                 object
education                                object
effective_to_date                datetime64[ns]
employment_status                        object
gender                                   object
income                                    int64
location_code                            object
marital_status                           object
monthly_premium_auto                      int64
months_since_last_claim                   int64
months_since_policy_inception             int64
number_of_open_complaints                 int64
number_of_policies                        int64
policy_type                              object
policy                                   object
renew_offer_type                         object
sales_channel                           

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   customer                       10910 non-null  object        
 1   state                          10910 non-null  object        
 2   customer_lifetime_value        10910 non-null  float64       
 3   response                       10910 non-null  object        
 4   coverage                       10910 non-null  object        
 5   education                      10910 non-null  object        
 6   effective_to_date              10910 non-null  datetime64[ns]
 7   employment_status              10910 non-null  object        
 8   gender                         10910 non-null  object        
 9   income                         10910 non-null  int64         
 10  location_code                  10910 non-null  object        
 11  marital_status 

None

In [42]:
marketing_customer_df_clean.to_csv("marketing_customer_analysis_clean.csv", index = False)

In [43]:
!ls


README.md
__pycache__
cleanfunctions.py
lab-dw-data-structuring-and-combining.ipynb
lab-dw-data-structuring-and-combining_Minh.ipynb
marketing_customer_analysis_clean
marketing_customer_analysis_clean.csv


1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [38]:
#a summary table showing the total revenue for each sales channel (branch, call center, web, and mail)

np.round(marketing_customer_df_clean.pivot_table(index = 'sales_channel', values = 'customer_lifetime_value', aggfunc = sum),2)



customer_lifetime_value
sales_channel                         
Agent                      33057887.85
Branch                     24359201.21
Call Center                17364288.37
Web                        12697632.90

# insights
the output shows the channel which has the highest total revenue is from agent, followed by channel Branch, Call Center, and the least comes from Web


2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [39]:
# average customer lifetime value per gender and education level
print('The below table show average customer lifetime value per gender and education level: ')

np.round(marketing_customer_df_clean.pivot_table(index = ['gender', 'education'], values = 'customer_lifetime_value'),2)


The below table show average customer lifetime value per gender and education level: 


customer_lifetime_value
gender education                                    
F      Bachelor                              7874.27
       College                               7748.82
       Doctor                                7328.51
       High School or Below                  8675.22
       Master                                8157.05
M      Bachelor                              7703.60
       College                               8052.46
       Doctor                                7415.33
       High School or Below                  8149.69
       Master                                8168.83

# insights
the average customer lifetime values are quite similar among these groups. 
Another way to look at this is that it seems gender and education level has no effect/impact on the average customer lifetime value.
however, to be able to draw a solid conclusion, it should be statistically tested whether the small differences that we observed among these groups are significant 

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [45]:
# Your code goes here
# I want to check this dataset is from one year or from multiple years
check = pd.DataFrame()
check['year'] = marketing_customer_df_clean['effective_to_date'].dt.year
check.value_counts()
# output shows this is only from year 2011

# I make an extra column named 'effective_to_month' which extracts month from 'effective_to_date'
marketing_customer_df_clean['effective_to_month'] = marketing_customer_df_clean['effective_to_date'].dt.month

# table with total complaints by policy type and month
total_complaints_by_policy_type_and_month = marketing_customer_df_clean.groupby(['policy_type','effective_to_month' ]).agg({'number_of_open_complaints':'sum'})
display(total_complaints_by_policy_type_and_month.head())

# test to change result into long format

test = total_complaints_by_policy_type_and_month.reset_index()
display(test.head())

pd.melt(test, id_vars = ['policy_type'], value_vars = ['effective_to_month','number_of_open_complaints'] )

number_of_open_complaints
policy_type    effective_to_month                           
Corporate Auto 1                                         415
               2                                         361
Personal Auto  1                                        1635
               2                                        1363
Special Auto   1                                          84

policy_type  effective_to_month  number_of_open_complaints
0  Corporate Auto                   1                        415
1  Corporate Auto                   2                        361
2   Personal Auto                   1                       1635
3   Personal Auto                   2                       1363
4    Special Auto                   1                         84

policy_type                   variable  value
0   Corporate Auto         effective_to_month      1
1   Corporate Auto         effective_to_month      2
2    Personal Auto         effective_to_month      1
3    Personal Auto         effective_to_month      2
4     Special Auto         effective_to_month      1
5     Special Auto         effective_to_month      2
6   Corporate Auto  number_of_open_complaints    415
7   Corporate Auto  number_of_open_complaints    361
8    Personal Auto  number_of_open_complaints   1635
9    Personal Auto  number_of_open_complaints   1363
10    Special Auto  number_of_open_complaints     84
11    Special Auto  number_of_open_complaints     91

In [41]:
check = pd.DataFrame()
check['year'] = marketing_customer_df_clean['effective_to_date'].dt.year
check.value_counts()
check['month'] = marketing_customer_df_clean['effective_to_date'].dt.month
check.value_counts()

year  month
2011  1        5818
      2        5092
dtype: int64